develop/blockchains/custom-blockchains/benchmarking.md

```diff
@@ -1,4 +1,191 @@
+
+Create a `benchmarking.rs` file in your pallet's `src/`. You may copy the barebones template below to get started: 
+
+!!!note "This example is from the pallet template"
+ Take a look at the repository in the [`polkadot-sdk-parachain-template`](https://github.com/paritytech/polkadot-sdk-parachain-template/tree/master/pallets){target=_blank}
+
+```rust
+//! Benchmarking setup for pallet-template
+#![cfg(feature = "runtime-benchmarks")]
+
+use super::*;
+use frame_benchmarking::v2::*;
+
+#[benchmarks]
+mod benchmarks {
+    use super::*;
+ #[cfg(test)]
+    use crate::pallet::Pallet as Template;
+    use frame_system::RawOrigin;
+
+ #[benchmark]
+    fn do_something() {
+        let caller: T::AccountId = whitelisted_caller();
+ #[extrinsic_call]
+ do_something(RawOrigin::Signed(caller), 100);
+ assert_eq!(Something::<T>::get().map(|v| v.block_number), Some(100u32.into()));
+ }
+
+ impl_benchmark_test_suite!(Template, crate::mock::new_test_ext(), crate::mock::Test);
+}
+```
+
+The function `do_something` is a placeholder; you must write your own function that tested your extrinsic. Similar to writing unit tests, you have access to the mock runtime, and can use functions such as `whitelisted_caller()` (an account whitelisted for DB reads/writes) to sign transactions in the benchmarking context.
+
+There are a couple practices to notice: 
+
+- The `#[extrinsic_call]` macro is used when calling the extrinsic itself. This macro is a required part of a benchmarking function, [see the Rust docs](https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/v2/index.html#extrinsic_call-and-block){target=_blank} for more clarification
+- The `assert_eq` expression ensures that the extrinsic is working properly within the benchmark context
+
+## Adding Benchmarks to Runtime
+
+The last step before running the benchmarking tool is to ensure that the benchmarks are configured as part of your runtime.
+
+Create another file in the node's directory called `benchmarks.rs`. This is where the pallets you wish to benchmark will be registered. This file should contain the following macro, which registers all pallets for benchmarking, as well as their respective configurations: 
+
+```rust
+frame_benchmarking::define_benchmarks!(
+ [frame_system, SystemBench::<Runtime>]
+ [pallet_balances, Balances]
+ [pallet_session, SessionBench::<Runtime>]
+ [pallet_timestamp, Timestamp]
+ [pallet_message_queue, MessageQueue]
+ [pallet_sudo, Sudo]
+ [pallet_collator_selection, CollatorSelection]
+ [cumulus_pallet_parachain_system, ParachainSystem]
+ [cumulus_pallet_xcmp_queue, XcmpQueue]
+);
+```
+
+For example, if the pallet named `pallet_template` is ready to be benchmarked, it may be added as follows: 
+
+```rust
+frame_benchmarking::define_benchmarks!(
+    // ...
+ [pallet_template, Template]
+    // ...
+);
+```
+
+!!!warning 
+ If the pallet isn't included in the `define_benchmarks!` macro, the CLI will not be able to access it and benchmark it later.
```

Comment:
Check other admonitions because they are not being rendered properly. The actual content needs to be indented with 4 spaces like here
```suggestion

!!!warning 
    If the pallet isn't included in the `define_benchmarks!` macro, the CLI will not be able to access it and benchmark it later.
```